scripts/irods/test/test_all_rules.py

```diff
@@ -1651,3 +1650,309 @@ def test_chksum_true_forceChksum_1_verifyChksum_1__issue_5927(self):
+            "msi_json_size(*h, '/', *ignored);",
+            "msi_json_type(*h, '/', *ignored);",
+            "msi_json_value(*h, '/', *ignored);"
+        ]
+
+        for msi_call in msi_calls:
+            rule_text = textwrap.dedent('''
+                *json_string = '{{"x": 100}}';
+                msiStrlen(*json_string, *size);
+
+                msi_json_parse(*json_string, int(*size), *h);
+                {0}
+            '''.format(msi_call))
+
+            self.user.assert_icommand(['irule', '-r', rep_name, rule_text, 'null', 'ruleExecOut'], 'STDERR', ['DOES_NOT_EXIST'])
+
+    @unittest.skipUnless(plugin_name == 'irods_rule_engine_plugin-irods_rule_language', 'Only applicable to the NREP')
+    def test_msi_json_compare__issue_6968(self):
+        json_object = json.dumps({
+            'p1': 'alice',
+            'p2': 'bob',
+            'data': {'values': [1, 2, 3]}
+        })
+
+        def do_test(json_ptr_1, json_ptr_2, cmp_result):
+            rule_text = textwrap.dedent('''
+                msiStrlen('{0}', *size);
+
+                msi_json_parse('{0}', int(*size), *h);
+                msi_json_compare(*h, '{1}', *h, '{2}', *cmp_result);
+                msi_json_free(*h);
+
+                writeLine('stdout', 'cmp_result=[*cmp_result]');
+            '''.format(json_object, json_ptr_1, json_ptr_2))
+
+            rep_name = 'irods_rule_engine_plugin-irods_rule_language-instance'
+            self.user.assert_icommand(['irule', '-r', rep_name, rule_text, 'null', 'ruleExecOut'], 'STDOUT', ['cmp_result=[{0}]'.format(cmp_result)])
+
+        do_test(   '',    '', 'equivalent')
+        do_test('/p1', '/p2', 'less')
+        do_test('/p2', '/p1', 'greater')
+        #do_test(   '', '/p1', 'unordered') # TODO Requires the ability to generate NaN.
+
+    @unittest.skipUnless(plugin_name == 'irods_rule_engine_plugin-irods_rule_language', 'Only applicable to the NREP')
+    def test_msi_json_contains__issue_6968(self):
+        rule_text = textwrap.dedent('''
+            *json = '{"name": "john"}';
+            msiStrlen(*json, *size);
+
+            msi_json_parse(*json, int(*size), *h);
+            msi_json_contains(*h, '/name', *result_1);
+            msi_json_contains(*h, '/nope', *result_2);
+            msi_json_free(*h);
+
+            writeLine('stdout', 'result_1=[*result_1]');
+            writeLine('stdout', 'result_2=[*result_2]');
+        ''')
+
+        rep_name = 'irods_rule_engine_plugin-irods_rule_language-instance'
+        expected_output = ['result_1=[1]', 'result_2=[0]']
+        self.user.assert_icommand(['irule', '-r', rep_name, rule_text, 'null', 'ruleExecOut'], 'STDOUT', expected_output)
+
+    @unittest.skipUnless(plugin_name == 'irods_rule_engine_plugin-irods_rule_language', 'Only applicable to the NREP')
+    def test_msi_json_dump__issue_6968(self):
+        json_string = json.dumps({
+            'fname': 'john',
+            'lname': 'doe'
+        }, separators=(',', ':'), sort_keys=True)
```

Comment:
Had to add `sort_keys=True` here.

This wasn't necessary for the Python 3 version. I'll double check though.

Reply:
That is strange. It seems like if you define them in alphabetical order that's just how they would appear...

Or is the problem more with the expected output? Hmm...

Reply:
the json spec explicitly says it's not sorted... as a format.

Reply:
So the test in main/4-3-stable has a 50% chance of failing since `sort_keys` is `False` (default)?

Reply:
Python3.7+ remembers sort order by default.

Reply:
Marking as resolved.